akka-docs/src/main/paradox/scala/stream/stream-error.md

```diff
@@ -1,18 +1,134 @@
-strategies, but the semantics have been adapted to the domain of stream processing.
+When a stage in a stream fails this will normally lead to the entire stream being torn down.
+Each of the stages downstream gets informed about the failure and each upstream stage sees a cancellation.
 
-@@@ warning
+In many cases you may want to avoid complete stream failure, this can be done in a few different ways:
 
-*ZipWith*, *GraphStage* junction, *ActorPublisher* source and *ActorSubscriber* sink
-components do not honour the supervision strategy attribute yet.
+ * `recover` to emit a final element then complete the stream normally on upstream failure 
+ * `recoverWithRetries` to create a new upstream and start consuming from that on failure
+ * Restarting sections of the stream after a backoff
+ * Using a supervision strategy for stages that support it
+ 
+In addition to these built in tools for error handling, a common pattern is to wrap the stream 
+inside an actor, and have the actor restart the entire stream on failure.
+ 
+## Recover
+
+`recover` allows you to emit a final element and then complete the stream on an upstream failure.
+Deciding which exceptions should be recovered is done through a `PartialFunction`. If an exception
+does not have a @scala[matching case] @java[match defined] the stream is failed. 
+
+Recovering can be useful if you want to gracefully complete a stream on failure while letting 
+downstream know that there was a failure.
+
+Scala
+:   @@snip [FlowErrorDocSpec.scala]($code$/scala/docs/stream/FlowErrorDocSpec.scala) { #recover }
+
+Java
+:   @@snip [FlowErrorDocTest.java]($code$/java/jdocs/stream/FlowErrorDocTest.java) { #recover }
+
+This will output:
+
+Scala
+:   @@snip [FlowErrorDocSpec.scala]($code$/scala/docs/stream/FlowErrorDocSpec.scala) { #recover-output }
+
+Java
+:   @@snip [FlowErrorDocTest.java]($code$/java/jdocs/stream/FlowErrorDocTest.java) { #recover-output }
 
-@@@
+
+## Recover with retries
+
+`recoverWithRetries` allows you to put a new upstream in place of the failed one, recovering 
+stream failures up to a specified maximum number of times. 
+
+Deciding which exceptions should be recovered is done through a `PartialFunction`. If an exception
+does not have a @scala[matching case] @java[match defined] the stream is failed.
+
+Scala
+:   @@snip [FlowErrorDocSpec.scala]($code$/scala/docs/stream/FlowErrorDocSpec.scala) { #recoverWithRetries }
+
+Java
+:   @@snip [FlowErrorDocTest.java]($code$/java/jdocs/stream/FlowErrorDocTest.java) { #recoverWithRetries }
+
+This will output:
+
+
+Scala
+:   @@snip [FlowErrorDocSpec.scala]($code$/scala/docs/stream/FlowErrorDocSpec.scala) { #recoverWithRetries-output }
+
+Java
+:   @@snip [FlowErrorDocTest.java]($code$/java/jdocs/stream/FlowErrorDocTest.java) { #recoverWithRetries-output }
+
+
+## Delayed restarts with a backoff stage
+
+Just as Akka provides the @ref:[backoff supervision pattern for actors](../general/supervision.md#backoff-supervisor), Akka streams
```

Comment:
IIRC it's different ways to write the same - http://developer.lightbend.com/docs/paradox/latest/features/linking.html suggests `@ref:[` though - perhaps `@ref[` will not be recognized as a link by github? If so we should prefer `@ref:[` ;).